docs/rfc/rfc-018-bls-agg-exploration.md

```diff
@@ -0,0 +1,400 @@
+performance testing.
+
+Each validator will still need to receive vote extension data
+from the peer validators in order for consensus to proceed. As a result, any
+advantage gained by aggregating signatures across the vote message will be
+nullified as a result of the addition of vote extensions.
+
+#### Block Creation
+
+When creating a block, the proposer may create a small set of short
+multi-signatures and attach these to the block instead of including one
+signature per validator.
+
+#### Block Verification
+
+Verification of blocks would not verify a set of many signatures. Verification
+would instead check the single multi-signature using the public keys stored
+by the validator. Currently, we verify each validator signature using
+the public key associated with that validator.
+
+#### IBC Relaying
+
+IBC would no longer need to transmit a large set of signatures when
+updating state. These state updates do not happen for every IBC packet, only
+when changing an IBC light client's view of the counterparty chain's state.
+General [packets][ibc-packet] only contain enough information to correctly
+route the data to the counterparty chain.
+
+IBC does persist commit signatures to the chain in these `MsgUpdateClient`
+message when updating state. This message would no longer need the full set
+of unique signatures and would instead only need one signature for all of the
+data in the header.
+
+Adding BLS signatures would create a new signature type that must be
+understood by the IBC module and by the relayers. For some operations, such
+as state updates, the set of data written into the chain and received by the
+IBC module would be slightly smaller.
+
+## Discussion
+
+### What are the proposed benefits to aggregated signatures?
+
+#### Reduce Block Size
+
+At the moment, the commit size contains a 512 bit signature for each validator
+that voted for the block. For the cosmos hub, which has 175 validators in the
+validator set, this amounts to about 11kb per block. That is around 113 GB over
+the lifetime of the chain's 10118561 blocks. (Note, the hub has increased the
+number of validators in the set over time so the signature size over the history
+of the chain is likely less than that value). 
+
+Signature aggregation would only produce two signatures for the entire block.
+One for the yays and one for the nays. Each BLS aggregated signature is 48
+bytes, per the [IETF standard of BLS signatures][bls-ietf-ecdsa-compare].
+Over the lifetime of the same cosmos hub chain, that would amount to about 1
+GB, a savings of 112 GB. While that is a large factor of reduction it's worth
+bearing in mind that, at [GCP's cost][gcp-storage-pricing] of $.026 USD per GB,
+that is a total savings of around $2.50 per month.
+
+
+
+
+#### Reduce Signature Verification Creation and Verification Time
+
+From the [IETF draft standard on BLS Signatures][bls-ietf], BLS signatures can be
+created in 370 microseconds and verified in 2700 microseconds. Our current
```

Comment:
I'm not seeing these numbers in that doc. There are a couple of other benchmark links, but I didn't see these specific numbers. Which doc should we be looking at?

Won't the time to create the signature depend on the size of the signing set? (Even if the output size is fixed)

Reply:
Relatedly: Are these numbers from a specific reference implementation? If we want to compare these to the Go values for our own Ed25519 implementation, we'll probably need to compare to a Go BLS build.

Reply:
> I'm not seeing these numbers in that doc. There are a couple of other benchmark links, but I didn't see these specific numbers. Which doc should we be looking at?

Those values should be in [section 1.1](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-04#section-1.1) of the ietf draft on BLS, unless I've linked the wrong document here?

```
The following comparison assumes BLS signatures with curve BLS12-381, targeting 128 bits security. For 128 bits security, ECDSA takes 37 and 79 micro-seconds to sign and verify a signature on a typical laptop.  In comparison, for the same level of security, BLS takes 370 and 2700 micro-seconds to sign and verify a signature.
```

> Relatedly: Are these numbers from a specific reference implementation? If we want to compare these to the Go values for our own Ed25519 implementation, we'll probably need to compare to a Go BLS build.

The doc is pretty unclear on that. [I performed a benchmark](https://github.com/williambanfield/blst/blame/bench/bindings/go/PERFORMANCE.md#L9) using `blst` and linked the results below in the doc, so hopefully those are valuable.

Reply:
D'oh. I grepped for "microseconds" and "µs" but not "micro-seconds" haha.

That said—my guess is that their reference implementation is probably in C, given the usual practices of IETF.